Revert "functionalization: add support for zero_()"

bdhirsh opened the pull request:
Reverting https://github.com/pytorch/pytorch/pull/75913 as it broke torchvision (see comment at https://github.com/pytorch/pytorch/pull/75913#issuecomment-1109908549)

Diagnose: The following code now fails but used to work:
```
import torch

def foo(a):
    b = torch.zeros_like(a, dtype=torch.bool)
    return b

a = torch.ones(2, requires_grad=True)
sfoo = torch.jit.script(foo)
sfoo(a)
```

Why? The reverted PR added an autograd formula for `zeros_like()`, which used to be a `CompositeImplicitAutograd` op.

Unfortunately, that changed the behavior of zeros_like as follows.  The `*_like` ops "work" with autograd, but they sever the autograd graph.
```
>>> a = torch.ones(2, requires_grad=True)
>>> b = torch.zeros_like(a)
>>> b.requires_grad
False
>>> b.is_leaf
True
```

That makes code like `torch.zeros_like(a, dtype=torch.bool)` valid even if `a` requires_grad: if the requires_grad-ness were propagated, autograd would throw an error that you can't use autograd with `bool` tensors.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76377 Add back: "functionalization: add support for zero_()"
* **#76375 Revert "functionalization: add support for zero_()"**

This reverts commit 7d44b3675bafdfbd59e6c81734ca3febd771dd7b.

bdhirsh committed:
Revert "functionalization: add support for zero_()"

This reverts commit 7d44b3675bafdfbd59e6c81734ca3febd771dd7b.

[ghstack-poisoned]

datumbox reviewed (approved):
Stamping to unblock the revert.

bdhirsh committed:
Update on "Revert "functionalization: add support for zero_()""

Reverting https://github.com/pytorch/pytorch/pull/75913 as it broke torchvision (see comment at https://github.com/pytorch/pytorch/pull/75913#issuecomment-1109908549)

Diagnose: The following code now fails but used to work:
```
import torch

def foo(a):
    b = torch.zeros_like(a, dtype=torch.bool)
    return b

a = torch.ones(2, requires_grad=True)
sfoo = torch.jit.script(foo)
sfoo(a)
```

Why? The reverted PR added an autograd formula for `zeros_like()`, which used to be a `CompositeImplicitAutograd` op.

Unfortunately, that changed the behavior of zeros_like as follows.  The `*_like` ops "work" with autograd, but they sever the autograd graph.
```
>>> a = torch.ones(2, requires_grad=True)
>>> b = torch.zeros_like(a)
>>> b.requires_grad
False
>>> b.is_leaf
True
```

That makes code like `torch.zeros_like(a, dtype=torch.bool)` valid even if `a` requires_grad: if the requires_grad-ness were propagated, autograd would throw an error that you can't use autograd with `bool` tensors.


This reverts commit 7d44b3675bafdfbd59e6c81734ca3febd771dd7b.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

bdhirsh commented:
@pytorchbot merge this please
